src/swap/SwapScreen.tsx

```diff
@@ -62,7 +63,126 @@ const DEFAULT_SWAP_AMOUNT: SwapAmount = {
+  switchedToNetworkId: NetworkId | null
+}
+
+function getInitialState(fromTokenId?: string): SwapState {
+  return {
+    fromTokenId,
+    toTokenId: undefined,
+    swapAmount: DEFAULT_SWAP_AMOUNT,
+    updatedField: Field.FROM,
+    selectingToken: null,
+    confirmingSwap: false,
+    startedSwapId: null,
+    switchedToNetworkId: null,
+  }
+}
+
+const swapSlice = createSlice({
+  name: 'swapSlice',
+  initialState: getInitialState,
+  reducers: {
+    changeAmount: (state, action: PayloadAction<{ field: Field; value: string }>) => {
+      const { field, value } = action.payload
+      state.confirmingSwap = false
+      state.startedSwapId = null
+      state.updatedField = field
+      if (!value) {
+        state.swapAmount = DEFAULT_SWAP_AMOUNT
+        return
+      }
+      // Regex to match only numbers and one decimal separator
+      const sanitizedValue = value.match(/^(?:\d+[.,]?\d*|[.,]\d*|[.,])$/)?.join('')
+      if (!sanitizedValue) {
+        return
+      }
+      state.swapAmount[field] = sanitizedValue
+    },
+    chooseMaxFromAmount: (state, action: PayloadAction<{ fromTokenBalance: BigNumber }>) => {
+      const { fromTokenBalance } = action.payload
+      state.confirmingSwap = false
+      state.startedSwapId = null
+      state.updatedField = Field.FROM
+      // We try the current balance first, and we will prompt the user if it's too high
+      state.swapAmount[Field.FROM] = fromTokenBalance.toFormat({
+        decimalSeparator: getNumberFormatSettings().decimalSeparator,
+      })
+    },
+    startSelectToken: (state, action: PayloadAction<{ fieldType: Field }>) => {
+      state.selectingToken = action.payload.fieldType
+      state.confirmingSwap = false
+    },
+    selectTokens: (
+      state,
+      action: PayloadAction<{
+        fromTokenId: string | undefined
+        toTokenId: string | undefined
+        switchedToNetworkId: NetworkId | null
+      }>
+    ) => {
+      const { fromTokenId, toTokenId, switchedToNetworkId } = action.payload
+      state.confirmingSwap = false
+      if (fromTokenId !== state.fromTokenId || toTokenId !== state.toTokenId) {
+        state.startedSwapId = null
+      }
+      state.fromTokenId = fromTokenId
+      state.toTokenId = toTokenId
+      state.switchedToNetworkId = switchedToNetworkId
+    },
+    quoteUpdated: (state, action: PayloadAction<{ quote: QuoteResult | null }>) => {
+      const { quote } = action.payload
+      const { updatedField } = state
+      state.confirmingSwap = false
+      const otherField = updatedField === Field.FROM ? Field.TO : Field.FROM
+      if (!quote) {
+        state.swapAmount[otherField] = ''
+        return
+      }
+
+      const { decimalSeparator } = getNumberFormatSettings()
+      const parsedAmount = parseInputAmount(state.swapAmount[updatedField], decimalSeparator)
+
+      const newAmount = parsedAmount.multipliedBy(
+        new BigNumber(quote.price).pow(updatedField === Field.FROM ? 1 : -1)
```

Comment:
why we need the trick `.pow(-1)`?

Reply:
It's because the price doesn't mean the same thing if it's buy vs sell flow.
I'm not sure if we could have done it differently / easier to understand.
But It'll be removed soon though.